File: 1289.minimum-falling-path-sum-ii.py
#
# @lc app=leetcode id=1289 lang=python3
#
# [1289] Minimum Falling Path Sum II
#

# @lc code=start
from typing import List


class Solution:
    '''
    dp[i][j] = minimum of all falling path which ends at row i and column j
    Then we can just iterate through the rows from top to bottom    
    '''
    def minFallingPathSum(self, grid: List[List[int]]) -> int:
        n = len(grid)
        if n == 1:
            return grid[0][0]
        for i in range(n):
            dp_new = []
            if i == 0:
                for j in range(n):
                    dp_new.append([grid[i][j], j])
            else:
                for j in range(n):
                    if j != dp[0][1]:
                        dp_new.append([grid[i][j] + dp[0][0], j])
                    else:
                        dp_new.append([grid[i][j] + dp[1][0], j])
            dp = dp_new[:]
            dp.sort()
        return dp[0][0]
                    
# @lc code=end

File: 1289.minimum-falling-path-sum-ii.cs
/*
 * @lc app=leetcode id=1289 lang=csharp
 *
 * [1289] Minimum Falling Path Sum II
 */

// @lc code=start
using System;
using System.Collections.Generic;

public class Solution {
    /*
    dp[i][j] = minimum of all falling path which ends at row i and column j
    Then we can just iterate through the rows from top to bottom    
    */
    public int MinFallingPathSum(int[][] grid) {
        int n = grid.Length;
        if (n == 1) {
            return grid[0][0];
        }
        
        List<int[]> dp = new List<int[]>();
        for (int i = 0; i < n; i++) {
            List<int[]> dpNew = new List<int[]>();
            if (i == 0) {
                for (int j = 0; j < n; j++) {
                    dpNew.Add(new int[]{grid[i][j], j});
                }
            } else {
                for (int j = 0; j < n; j++) {
                    if (j != dp[0][1]) {
                        dpNew.Add(new int[]{grid[i][j] + dp[0][0], j});
                    } else {
                        dpNew.Add(new int[]{grid[i][j] + dp[1][0], j});
                    }
                }
            }
            dp = dpNew;
            dp.Sort((x, y) => x[0].CompareTo(y[0]));
        }
        
        return dp[0][0];
    }
}

// @lc code=end

